### ISSUE TITLE
simulate module

### ISSUE THREAD

**Issue by Jammy2211** [Jammy2211/PyAutoLens#23]
I have written a very basic simulate module, which takes an image and adds PSF blurring / noise. However, the simulate module will be a key part of AutoLens, and will requite a lot more effort!

For simulating images,  basically two things are needed:

- The code to simulate instrumental effects (e.g. read noise, poisson noise, PSF, background sky, etc.)
- An interface that allows the user to control these effects whilst not being bogged down in hard to understanding parameters.

On point 1 - The majority of effects we might want to simulate already have existing functions from libraries like AstroPy. In truth, other than the PSF, background noise and Poisson noise we probably won't want to worry about another effect for a while. I guess correlated noise due to image dithering would be the next, which goes hand-in-hand with multi image analysis and thus should be postponed until then.

On point 2 - To produce an image with appropriate signal-to-noise, I had to both change the background level of noise and the exposure time. Sometimes, the source brightness, lens light brightness and lensing magnification will also be parameters one is changing. For a user messing with parameters it becomes confusing whether the images they are producing are sensible or realistic.

When generating simulated images, users will want to interact with the code and generate images follwing two very different approaches:

- As physical as possible, using realistic magnitudes for the lens / source galaxies and realistic mass models, as well as physical parameters for the instrument. Thus, we probably want some classes which default to observed data-set results and instruments.

- Exploratory - e.g. where you ask questions like 'what does a broader PSF or lower S/N do to my results'. Here, you just want the images to come out 'looking right' but in a way you can change whatever you're interested in. For the problem above, we'd have want an automated routine that produces an image to a specified S/N.

Furthermore, it worth remember that we will want to tie the simulate module directly to the analysis modules, so one can swiftly simulate images and then see how fits behave.

**Comment by Jammy2211** [comment]
I have cleaned up the simulate module a bit - the effects are now their own classes and can more easily be switched on or off. This will probably still be completely refactored, but makes editing more straightforward for when it comes to sorting this.

**Comment by rhayes777** [comment]
In some cases these classes seem that they could be replaced with simple functions.

Some instance attributes are being associated with the instance by a function other than __init__. Those should probably be converted to properties or calculated at instantiation, depending on how often they are likely to be called.

**Comment by Jammy2211** [comment]
Okay, so looking at your simulate classmethod, there were a few errors introduced (but you made it a lot simpler and thus easy to spot!)

Firstly, I've changed the way simulate works to make it more 'physical'. Rather than adding the background at the end (as Gaussian noise), I've included the background sky in the initial image such that it is included via the Poisson noise term. The motivation here is that for a non-constant background sky this will now behave physical (albeit lets just assume for now the background sky is constant).

The code can thus simulate 3 effects, which are in order:

- Add background sky.
- Simulate optics via PSF.
- Add Poisson noise (including background sky)
- (subtract background sky, as AutoLens uses background subtracted images. This uses the input background sky map).

Previously, the method treated each effect independently. For example, the Poisson noise was added to the input array, NOT the input array after PSF blurring. I've rectified this, and added tests which correctly test the behaviour for multiple features being on / off.

As you'll see, the code is a bit messy now - mainly because I dont know how to use properties in a classmethod. However, it is still quite each to pick apart what is happening!

There will be more effects to add in future, but I think these 3 will suffice for making sure our simulate module works.
